src/MySqlConnector/MySqlClient/Types/TypeMapper.cs

```diff
@@ -0,0 +1,141 @@
+
+			// date/time
+			var typeDateTime       = AddDbTypeMapping(new DbTypeMapping(typeof(DateTime), new []{DbType.DateTime, DbType.Date, DbType.DateTime2}));
+			var typeTime           = AddDbTypeMapping(new DbTypeMapping(typeof(TimeSpan), new []{DbType.Time}));
+			var typeDateTimeOffset = AddDbTypeMapping(new DbTypeMapping(typeof(DateTimeOffset), new []{DbType.DateTimeOffset}));
+			AddColumnTypeMapping(new ColumnTypeMapping("datetime",  typeDateTime,       new []{ColumnType.DateTime}));
+			AddColumnTypeMapping(new ColumnTypeMapping("date",      typeDateTime,       new []{ColumnType.Date}));
+			AddColumnTypeMapping(new ColumnTypeMapping("time",      typeTime,           new []{ColumnType.Time}));
+			AddColumnTypeMapping(new ColumnTypeMapping("timestamp", typeDateTimeOffset, new []{ColumnType.Timestamp}));
+			AddColumnTypeMapping(new ColumnTypeMapping("year",      typeInt,            new []{ColumnType.Year}));
+
+			// guid
+			AddDbTypeMapping(new DbTypeMapping(typeof(Guid), new[]{DbType.Guid}, convert: o => Guid.Parse(Convert.ToString(o))));
+		}
+
+		private DbTypeMapping AddDbTypeMapping(DbTypeMapping dbTypeMapping)
+		{
+			m_dbTypeMappingsByClrType[dbTypeMapping.ClrType] = dbTypeMapping;
+
+			if (dbTypeMapping.DbTypes != null)
+				foreach (var dbType in dbTypeMapping.DbTypes)
+					m_dbTypeMappingsByDbType[dbType] = dbTypeMapping;
+
+			return dbTypeMapping;
+		}
+
+		private void AddColumnTypeMapping(ColumnTypeMapping columnTypeMapping)
+		{
+			m_columnTypeMappingLookup[columnTypeMapping.LookupKey] = columnTypeMapping;
+		}
+
+		internal DbTypeMapping GetDbTypeMapping(Type clrType)
+		{
+			DbTypeMapping dbTypeMapping;
+			m_dbTypeMappingsByClrType.TryGetValue(clrType, out dbTypeMapping);
+			return dbTypeMapping;
+		}
+
+		internal DbTypeMapping GetDbTypeMapping(DbType dbType)
+		{
+			DbTypeMapping dbTypeMapping;
+			m_dbTypeMappingsByDbType.TryGetValue(dbType, out dbTypeMapping);
+			return dbTypeMapping;
+		}
+
+		internal DbTypeMapping GetDbTypeMapping(string columnTypeName, bool unsigned=false, int length=0)
+		{
+			return GetColumnTypeMapping(columnTypeName, unsigned, length)?.DbTypeMapping;
+		}
+
+		internal ColumnTypeMapping GetColumnTypeMapping(string columnTypeName, bool unsigned=false, int length=0)
+		{
+			ColumnTypeMapping columnTypeMapping;
+			if (!m_columnTypeMappingLookup.TryGetValue(ColumnTypeMapping.CreateLookupKey(columnTypeName, unsigned, length), out columnTypeMapping) && length != 0)
+				m_columnTypeMappingLookup.TryGetValue(ColumnTypeMapping.CreateLookupKey(columnTypeName, unsigned, 0), out columnTypeMapping);
+			return columnTypeMapping;
+		}
+
+		private Dictionary<Type, DbTypeMapping> m_dbTypeMappingsByClrType = new Dictionary<Type, DbTypeMapping>();
+		private Dictionary<DbType, DbTypeMapping> m_dbTypeMappingsByDbType = new Dictionary<DbType, DbTypeMapping>();
+		private Dictionary<string, ColumnTypeMapping> m_columnTypeMappingLookup = new Dictionary<string, ColumnTypeMapping>();
```

Comment:
Ah I see, good tip.  Since everything runs through CreateLookupKey and this is private, I'll just call `ToLowerInvariant()` there (no need to double-tap)